cmd/snap-confine/snap-confine.c

```diff
@@ -131,8 +154,115 @@ static void sc_preserve_and_sanitize_process_state(sc_preserved_process_state *
 static void sc_restore_process_state(const sc_preserved_process_state *
 				     proc_state)
 {
+	/* Restore original umask */
 	umask(proc_state->orig_umask);
 	debug("umask restored to %#4o", proc_state->orig_umask);
+
+	/* Restore original current working directory.
+	 *
+	 * This part is more involved for the following reasons. While we hold an
+	 * O_PATH file descriptor that still points to the original working
+	 * directory, that directory may not be representable in the target mount
+	 * namespace. A quick example may be /custom that exists on the host but
+	 * not in the base snap of the application.
+	 *
+	 * Also consider when the path of the original working directory now
+	 * maps to a different inode we cannot use fchdir(2). One example of
+	 * that is the /tmp directory, which exists in both the host mount
+	 * namespace and the per-snap mount namespace but actually represents a
+	 * different directory.
+	 **/
+
+	/* Read the target of symlink at /proc/self/fd/<fd-of-orig-cwd> */
+	char fd_path[PATH_MAX];
+	char orig_cwd[PATH_MAX];
+	ssize_t nread;
+	/* If the original working directory cannot be used for whatever reason then
+	 * move the process to a special void directory. */
+	const char *sc_void_dir = "/var/lib/snapd/void";
+
+	sc_must_snprintf(fd_path, sizeof fd_path, "/proc/self/fd/%d",
+			 proc_state->orig_cwd_fd);
+	nread = readlink(fd_path, orig_cwd, sizeof orig_cwd);
+	if (nread < 0) {
+		die("cannot read symbolic link target %s", fd_path);
+	}
+	if (nread == sizeof orig_cwd) {
+		die("cannot fit symbolic link target %s", fd_path);
+	}
+
+	/* Open path corresponding to the original working directory in the
+	 * execution environment. This may normally fail if the path no longer
+	 * exists here, this is not a fatal error. It may also fail if we don't
+	 * have permissions to view that path, that is not a fatal error either. */
+	int inner_cwd_fd SC_CLEANUP(sc_cleanup_close) = -1;
+	inner_cwd_fd =
+	    open(orig_cwd, O_PATH | O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW);
+	if (inner_cwd_fd < 0) {
+		if (errno == EPERM || errno == EACCES || errno == ENOENT) {
+			debug
```

Comment:
Wow, indent really does unspeakable things to the code.

Reply:
Yes... I'm on the edge of using clang-format for snap-confine.c because this is just silly.

On top of that, it's not stable, run indent a few times and you get different output.